miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultBoxService.java

```diff
@@ -239,91 +234,106 @@ public BoxUse getUse(long id) throws IOException {
+        BoxPosition newPos = entry.getValue();
 
-        if (previousOccupant != null && newOccupant.getId().equals(previousOccupant.getId())) {
+        if (managedPos != null && newPos.getBoxableId().equals(managedPos.getBoxableId())) {
           // Unchanged
           continue;
         }
         if (message.length() > 0) {
           message.append("\n");
         }
 
-        BoxableView oldOccupant = oldOccupants.get(newOccupant.getId());
+        BoxableView oldOccupant = oldOccupants.get(newPos.getBoxableId());
         if (oldOccupant.getBoxId() != null) {
           if (oldOccupant.getBoxId().longValue() == box.getId()) {
             // Moved within same box
             message.append(String.format("Relocated %s (%s) from %s to %s", oldOccupant.getAlias(), oldOccupant.getName(),
                 oldOccupant.getBoxPosition(), entry.getKey()));
+            movedWithinBox.add(newPos);
           } else {
             // Moved from a different box
             message.append(String.format("Moved %s (%s) from %s (%s) to %s", oldOccupant.getAlias(), oldOccupant.getName(),
                 oldOccupant.getBoxAlias(), oldOccupant.getBoxName(), entry.getKey()));
-
-            Box oldHome = boxStore.get(oldOccupant.getBoxId());
-            String oldHomeMessage = String.format("Moved %s (%s) to %s (%s)", oldOccupant.getAlias(), oldOccupant.getName(),
-                original.getAlias(), original.getName());
-            addBoxContentsChangeLog(oldHome, oldHomeMessage);
+            movedFromOtherBoxes.add(oldOccupant);
           }
-          boxStore.removeBoxableFromBox(oldOccupant);
         } else {
           message.append(String.format("Added %s (%s) to %s", oldOccupant.getAlias(), oldOccupant.getName(), entry.getKey()));
         }
       }
 
       // Process removals
-      for (Map.Entry<String, BoxableView> entry : original.getBoxables().entrySet()) {
-        if (box.getBoxables().keySet().contains(entry.getKey()) || handled.contains(entry.getValue().getId())) {
+      List<BoxableId> removedIds = new ArrayList<>();
+      List<BoxableId> movedWithinBoxIds = movedWithinBox.stream().map(BoxPosition::getBoxableId).collect(Collectors.toList());
+      for (Map.Entry<String, BoxPosition> entry : managed.getBoxPositions().entrySet()) {
+        if (box.getBoxPositions().keySet().contains(entry.getKey())) {
           // Already handled. Only checking for removals at this point
           continue;
         }
-        if (message.length() > 0) {
-          message.append("\n");
+        removedIds.add(entry.getValue().getBoxableId());
+      }
+      List<BoxableView> removed = boxStore.getBoxableViewsByIdList(removedIds);
+      for (BoxableView v : removed) {
+        if (!movedWithinBoxIds.contains(v.getId())) {
+          if (message.length() > 0) {
+            message.append("\n");
+          }
+          message.append(String.format("Removed %s (%s)", v.getAlias(), v.getName()));
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Define a constant instead of duplicating this literal "Removed %s (%s)" 3 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)